### PR TITLE
Mobile 2.0 API tests and System test

### PR DESCRIPTION
**Pull request by hrajput89**
#### Fixes #.

- [x] Ran `flake8`
- [x] Ran `run_repo_tests.sh`

#### Changes proposed in this pull request:

- API tests
- System tests

### CODE REVIEW

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_database.py [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+    ])
+    def test_database_create_exception(self, db_name, err_msg):
+        '''
+        @summary: Checking for the Exception handling in database create API
+        '''
+        if self.liteserv_platform != "android" and db_name == "":
+            pytest.skip("Test not applicable for ios")
+
+        if len(db_name) >= 128 and (self.liteserv_platform != "ios" or self.liteserv_platform != "android"):
+            pytest.skip("Test not supported on .net platfomrs")
```
Comment (raghusarangapani) [inline testsuites/CBLTester/CBL_Functional_tests/APITests/test_database.py]: s/platfomrs/platforms/

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_database.py [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+    def test_delete_doc(self, db_name, doc_id):
+        '''
+        @summary: Testing delete method of Database API
+        '''
+        if len(db_name) >= 128 and (self.liteserv_platform != "ios" or self.liteserv_platform != "android"):
+            pytest.skip("Test not supported on .net platfomrs")
+
+        db_name = db_name.lower()
+        db = self.db_obj.create(db_name)
+        # assert self.db_obj.getDocument(db, doc_id) == -1
```
Comment (raghusarangapani) [inline testsuites/CBLTester/CBL_Functional_tests/APITests/test_database.py]: Can this assert be removed?

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_dictionary.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+        # assert self.datatype.compare(value,
+        #                            self.dict_obj.getDate(content_dict,
+        #                                                   key))
+        assert self.datatype.compareDate(value, self.dict_obj.getDate(content_dict, key))
+
+    def test1_get_set_dictionary(self):
+        '''
+        @summary: Testing get and set Dictionary methods of Dictionary API
+        '''
+        # TODO ios gets back {} for self.datatype.hashMap()
```
Comment (raghusarangapani) [inline testsuites/CBLTester/CBL_Functional_tests/APITests/test_dictionary.py]: What is to be done here?

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_dictionary.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    ])
+    def test_get_set_double(self, key, value):
+        '''
+        @summary: Testing get and set Double methods of Dictionary API
+        '''
+        # TODO implementation does not work on ios
+        double_value = self.datatype.setDouble(float(value))
+        content_dict = self.dict_obj.create()
+        self.dict_obj.setDouble(content_dict, key, double_value)
+        # assert self.datatype.compare(double_value,
```
Comment (raghusarangapani) [inline testsuites/CBLTester/CBL_Functional_tests/APITests/test_dictionary.py]: Can these commented lines be removed?

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_dictionary.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+        (random_string(6), "{}".format(random.uniform(0, 1))),
+        (random_string(6), "{}".format(random.uniform(1, 10))),
+        (random_string(6), "{}".format(random.uniform(11, 100))),
+        (random_string(6), "{}".format(random.uniform(101, 1000)))
+    ])
+    def test_get_set_double(self, key, value):
+        '''
+        @summary: Testing get and set Double methods of Dictionary API
+        '''
+        # TODO implementation does not work on ios
```
Comment (raghusarangapani) [inline testsuites/CBLTester/CBL_Functional_tests/APITests/test_dictionary.py]: What needs to be done here?

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_dictionary.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+        (random_string(6), round(random.uniform(0, 1), 3)),
+        (random_string(6), round(random.uniform(1, 10), 3)),
+        (random_string(6), round(random.uniform(11, 100), 3)),
+        (random_string(6), round(random.uniform(101, 1000), 3))
+    ])
+    def test_get_set_float(self, key, value):
+        '''
+        @summary: Testing get and set Float methods of Dictionary API
+        '''
+        # TODO Precision issue with ios assert 978.5709 == 978.571
```
Comment (raghusarangapani) [inline testsuites/CBLTester/CBL_Functional_tests/APITests/test_dictionary.py]: Is this now fixed?

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_replication_api.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
+import pytest
```
Comment (raghusarangapani) [inline testsuites/CBLTester/CBL_Functional_tests/APITests/test_replication_api.py]: Is this needed? The replication functional tests cover most of the APIs

File: testsuites/CBLTester/System_test_multiple_device/conftest.py [hunk 8]
```diff
@@ -0,0 +1,290 @@
+import time
+import pytest
+
+from utilities.cluster_config_utils import persist_cluster_config_environment_prop
+from keywords.utils import log_info
+from keywords.utils import host_for_url
+from keywords.ClusterKeywords import ClusterKeywords
+from keywords.constants import CLUSTER_CONFIGS_DIR
+# from keywords.TestServerFactory import TestServerFactory
```
Comment (raghusarangapani) [inline testsuites/CBLTester/System_test_multiple_device/conftest.py]: Can this commented line be removed?

File: testsuites/CBLTester/System_test_multiple_device/conftest.py [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+    enable_sample_bucket = request.config.getoption("--enable-sample-bucket")
+    xattrs_enabled = request.config.getoption("--xattrs")
+    create_db_per_suite = request.config.getoption("--create-db-per-suite")
+    device_enabled = request.config.getoption("--device")
+    sg_ssl = request.config.getoption("--sg-ssl")
+    resume_cluster = request.config.getoption("--resume-cluster")
+    generator = request.config.getoption("--doc-generator")
+    no_db_delete = request.config.getoption("--no-db-delete")
+
+#     community_enabled = request.config.getoption("--community")
```
Comment (raghusarangapani) [inline testsuites/CBLTester/System_test_multiple_device/conftest.py]: I think these need to be uncommented

File: testsuites/CBLTester/System_test_multiple_device/conftest.py [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+        "sync_gateway_version": sync_gateway_version,
+        "target_admin_url": target_admin_url,
+        "enable_sample_bucket": enable_sample_bucket,
+        "cbl_db_list": cbl_db_list,
+        "db_name_list": db_name_list,
+        "base_url_list": base_url_list,
+        "query_obj_list": query_obj_list,
+        "sg_config": sg_config,
+        "db_obj_list": db_obj_list,
+        # "testserver_list": testserver_list,
```
Comment (raghusarangapani) [inline testsuites/CBLTester/System_test_multiple_device/conftest.py]: Can this commented line be removed?

File: testsuites/CBLTester/System_test_multiple_device/conftest.py [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+        "device_enabled": device_enabled,
+        "generator": generator,
+        "resume_cluster": resume_cluster
+    }
+
+    # Delete CBL database
+#     for db_name, testserver, base_url in zip(db_name_list,
```
Comment (raghusarangapani) [inline testsuites/CBLTester/System_test_multiple_device/conftest.py]: Can these commented lines be removed?

File: testsuites/CBLTester/System_test_multiple_device/conftest.py [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+            time.sleep(2)
+            db_obj.deleteDB(cbl_db)
+            time.sleep(1)
+
+        # Flush all the memory contents on the server app
+        log_info("Flushing server memory")
+        utils_obj = Utils(base_url)
+        utils_obj.flushMemory()
+        log_info("Stopping the test server")
+        # testserver.stop()
```
Comment (raghusarangapani) [inline testsuites/CBLTester/System_test_multiple_device/conftest.py]: This should be uncommented

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_dictionary.py (hunk 4 above)
Comment (hrajput89) [reply testsuites/CBLTester/CBL_Functional_tests/APITests/test_dictionary.py]: done

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_dictionary.py (hunk 5 above)
Comment (hrajput89) [reply testsuites/CBLTester/CBL_Functional_tests/APITests/test_dictionary.py]: Fixes already done. Removed TODO comments

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_dictionary.py (hunk 6 above)
Comment (hrajput89) [reply testsuites/CBLTester/CBL_Functional_tests/APITests/test_dictionary.py]: Yes. Removed Todo comment.

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_replication_api.py (hunk 7 above)
Comment (hrajput89) [reply testsuites/CBLTester/CBL_Functional_tests/APITests/test_replication_api.py]: Removed the API Tests.

File: testsuites/CBLTester/System_test_multiple_device/conftest.py (hunk 9 above)
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/conftest.py]: uncomment it, but haven't tested the option.

File: testsuites/CBLTester/System_test_multiple_device/conftest.py (hunk 10 above)
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/conftest.py]: Done.

File: testsuites/CBLTester/System_test_multiple_device/conftest.py (hunk 11 above)
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/conftest.py]: done

File: testsuites/CBLTester/System_test_multiple_device/conftest.py (hunk 12 above)
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/conftest.py]: done

**Comment by hrajput89** [comment]
Made changes what @raghusarangapani has asked for.

File: testsuites/CBLTester/CBL_Functional_tests/conftest.py [hunk 13]
```diff
@@ -0,0 +1,571 @@
...
+        log_info("Using document storage for sync meta data")
+        persist_cluster_config_environment_prop(cluster_config, 'xattrs_enabled', False)
+
+    if no_conflicts_enabled:
+        log_info("Running with no conflicts")
+        persist_cluster_config_environment_prop(cluster_config, 'no_conflicts_enabled', True)
+    else:
+        log_info("Running with allow conflicts")
+        persist_cluster_config_environment_prop(cluster_config, 'no_conflicts_enabled', False)
+
```
Comment (sridevi-15) [inline testsuites/CBLTester/CBL_Functional_tests/conftest.py]: you might need to add GSI changes which Raghu did in GSI PR

File: testsuites/CBLTester/CBL_Functional_tests/conftest.py [hunk 14]
```diff
@@ -0,0 +1,571 @@
...
+    cluster_utils = ClusterKeywords()
+    cluster_topology = cluster_utils.get_cluster_topology(cluster_config)
+    # c = cluster.Cluster(config=cluster_config)
+    # c.reset(sg_config_path=sg_config)
```
Comment (sridevi-15) [inline testsuites/CBLTester/CBL_Functional_tests/conftest.py]: remove comments?

File: testsuites/CBLTester/CBL_Functional_tests/conftest.py [hunk 15]
```diff
@@ -0,0 +1,571 @@
...
+    cbs_url = cluster_topology['couchbase_servers'][0]
+    cbs_ip = host_for_url(cbs_url)
+
+    # cluster = Cluster(cluster_config)
```
Comment (sridevi-15) [inline testsuites/CBLTester/CBL_Functional_tests/conftest.py]: remove comments

File: testsuites/CBLTester/System_test_multiple_device/conftest.py [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+                     help="Enable device if you want to run it on device", default=False)
+
+    parser.addoption("--community", action="store_true",
+                     help="If set, community edition will get picked up , default is enterprise", default=False)
+
+    parser.addoption("--sg-ssl",
+                     action="store_true",
+                     help="If set, will enable SSL communication between Sync Gateway and CBL")
+
+
```
Comment (sridevi-15) [inline testsuites/CBLTester/System_test_multiple_device/conftest.py]: can you also add GSI changes too which Raghu added in his PR?

File: testsuites/CBLTester/System_test_multiple_device/conftest.py [hunk 17]
```diff
@@ -0,0 +1,290 @@
...
+#         log_info("Downloading TestServer ...")
+#         # Download TestServer app
+#         testserver.download()
+#
+#         # Install TestServer app
+#         if device_enabled and platform == "ios":
+#             testserver.install_device()
+#         else:
+#             testserver.install()
+#         testserver_list.append(testserver)
```
Comment (sridevi-15) [inline testsuites/CBLTester/System_test_multiple_device/conftest.py]: is it by mistake you commented? Need download and install, right?

File: testsuites/CBLTester/System_test_multiple_device/conftest.py [hunk 18]
```diff
@@ -0,0 +1,290 @@
...
+    else:
+        log_info("Running with allow conflicts")
+        persist_cluster_config_environment_prop(cluster_config, 'no_conflicts_enabled', False)
+
+    if sg_ssl:
+        log_info("Enabling SSL on sync gateway")
+        persist_cluster_config_environment_prop(cluster_config, 'sync_gateway_ssl', True)
+        target_url = "wss://{}:4984/{}".format(sg_ip, sg_db)
+        target_admin_url = "wss://{}:4985/{}".format(sg_ip, sg_db)
+
```
Comment (sridevi-15) [inline testsuites/CBLTester/System_test_multiple_device/conftest.py]: Can you add GSI change from GSI PR?

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 19]
```diff
@@ -0,0 +1,283 @@
...
+    complex_doc
+from datetime import datetime, timedelta
+
+
+@pytest.mark.sanity
+@pytest.mark.listener
+@pytest.mark.replication
+@pytest.mark.parametrize("num_of_docs, num_of_updates, num_of_docs_to_update, num_of_docs_in_itr, num_of_doc_to_delete, num_of_docs_to_add, up_time", [
+    (1000000, 100, 100, 1000, 1000, 2000, 4 * 24 * 60),
+    # (500, 5, 10, 5, 10, 10, 1 * 20),
```
Comment (sridevi-15) [inline testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: uncomment this?

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 20]
```diff
@@ -0,0 +1,283 @@
...
+        _check_doc_count(db_obj_list, cbl_db_list)
+        count = db_obj_list[0].getCount(cbl_db_list[0])
+        itr_count = count / query_limit
+        if itr_count == 0:
+            itr_count = 1
+        for _ in range(itr_count):
+            existing_docs = db_obj_list[0].getDocIds(cbl_db_list[0], query_limit, query_offset)
+            doc_ids.update(existing_docs)
+            query_offset += query_limit
+        print "{} Docs in DB".format(len(doc_ids))
```
Comment (sridevi-15) [inline testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: you can remove or you can use log_info if you need this to be logged

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 21]
```diff
@@ -0,0 +1,283 @@
...
+    current_time = datetime.now()
+    running_time = current_time + timedelta(minutes=up_time)
+
+    _check_doc_count(db_obj_list, cbl_db_list)
+    x = 1
+    while(running_time - current_time > timedelta(0)):
+
+        print '*' * 20
+        print "Starting iteration no. {} of system testing".format(x)
+        print '*' * 20
```
Comment (sridevi-15) [inline testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: can you either remove print statements or replace with log_info?

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 22]
```diff
@@ -0,0 +1,283 @@
...
+        x += 1
+        ######################################
+        # Checking for doc update on SG side #
+        ######################################
+        docs_to_update = random.sample(doc_ids, num_of_docs_to_update)
+        sg_docs = sg_client.get_all_docs(url=sg_url, db=sg_db, auth=session)["rows"]
+        sg_docs = [doc for doc in sg_docs if doc["id"] in docs_to_update]
+        print "Updating {} docs on SG - {}".format(len(docs_to_update),
```
Comment (sridevi-15) [inline testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: can you either remove print statements or replace with log_info?

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 23]
```diff
@@ -0,0 +1,283 @@
...
+                                                   docs_to_update)
+        sg_client.update_docs(url=sg_url, db=sg_db, docs=sg_docs,
+                              number_updates=num_of_updates, auth=session, channels=channels_sg)
+
+        # Waiting until replicator finishes on all dbs
+        for repl_obj, repl, cbl_db, query in zip(replicator_obj_list,
+                                                 replicator_list,
+                                                 cbl_db_list,
+                                                 query_obj_list):
+            t = Thread(target=_replicaton_status_check, args=(repl_obj, repl))
```
Comment (sridevi-15) [inline testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: if it is just one method calling, do you need to use threads?  if you use multiple methods or blocks, then it is advantage to use threads, right?

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 24]
```diff
@@ -0,0 +1,283 @@
...
+        #######################################
+        docs_to_update = random.sample(doc_ids, num_of_docs_to_update)
+        i = 0
+        for db_obj, cbl_db, repl_obj, repl, query in zip(db_obj_list,
+                                                         cbl_db_list,
+                                                         replicator_obj_list,
+                                                         replicator_list,
+                                                         query_obj_list):
+            updates_per_db = len(docs_to_update) / len(db_obj_list)
+            print "Updating {} docs on {} db - {}".format(updates_per_db,
```
Comment (sridevi-15) [inline testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: can you either remove print statements or replace with log_info?

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 25]
```diff
@@ -0,0 +1,283 @@
...
+        sg_docs = sg_client.get_bulk_docs(url=sg_url, db=sg_db, doc_ids=list(docs_to_delete), auth=session)[0]
+        print "Deleting {} docs on SG - {}".format(len(docs_to_delete),
+                                                   docs_to_delete)
+        sg_client.delete_bulk_docs(url=sg_url, db=sg_db,
+                                   docs=sg_docs, auth=session)
+        for repl_obj, repl, cbl_db, query in zip(replicator_obj_list,
+                                                 replicator_list,
+                                                 cbl_db_list,
+                                                 query_obj_list):
+            t = Thread(target=_replicaton_status_check, args=(repl_obj, repl))
```
Comment (sridevi-15) [inline testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: same as above, having one method or one thread is necessary to use threads?

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 26]
```diff
@@ -0,0 +1,283 @@
...
+            i += docs_to_delete_per_db
+            time.sleep(5)
+            query.query_get_docs_limit_offset(cbl_db, limit=query_limit, offset=query_offset)
+
+            # Deleting docs will affect all dbs as they are synced with SG.
+            for repl_obj, repl, cbl_db, query in zip(replicator_obj_list,
+                                                     replicator_list,
+                                                     cbl_db_list,
+                                                     query_obj_list):
+                t = Thread(target=_replicaton_status_check, args=(repl_obj, repl))
```
Comment (sridevi-15) [inline testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: as you have start and join together for one thread, it runs only one thead, right?

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 27]
```diff
@@ -0,0 +1,283 @@
...
+                elif generator == "simple_user":
+                    data = simple_user()
+                else:
+                    data = simple()
+                data["channels"] = channels_sg
+                data["_id"] = doc_id
+                added_docs[doc_id] = data
+                new_doc_ids.append(doc_id)
+            doc_ids.update(new_doc_ids)
+            print "creating {} docs on {} - {}".format(len(docs_to_create),
```
Comment (sridevi-15) [inline testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: can you either remove print statements or replace with log_info?

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 28]
```diff
@@ -0,0 +1,283 @@
...
+            t.start()
+            t.join()
+            query.query_get_docs_limit_offset(cbl_db, limit=query_limit, offset=query_offset)
+            time.sleep(5)
+        doc_id_for_new_docs += num_of_docs_to_add
+        _check_doc_count(db_obj_list, cbl_db_list)
+
+        current_time = datetime.now()
+    # stopping replication
+    print "Test completed. Stopping Replicators"
```
Comment (sridevi-15) [inline testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: can you either remove print statements or replace with log_info?

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_database.py [hunk 29]
```diff
@@ -0,0 +1,368 @@
...
+        random_string(6, digit=True),
+        random_string(6).capitalize(),
+        random_string(6).upper(),
+    ])
+    def test_database_create(self, db_name):
+        '''
+        @summary: Testing Database constructor method of Database API
+        '''
+        if len(db_name) >= 128 and (self.liteserv_platform != "ios" or self.liteserv_platform != "android"):
+            pytest.skip("Test not supported on .net platforms")
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/CBL_Functional_tests/APITests/test_database.py]: added in skip message.

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_database.py [hunk 30]
```diff
@@ -0,0 +1,368 @@
...
+        random_string(6, digit=True),
+        random_string(6).capitalize(),
+        random_string(6).upper(),
+    ])
+    def test_database_close(self, db_name):
+        '''
+        @summary: Testing close method of Database API
+        '''
+        if len(db_name) >= 128 and (self.liteserv_platform != "ios" or self.liteserv_platform != "android"):
+            pytest.skip("Test not supported on .net platfomrs")
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/CBL_Functional_tests/APITests/test_database.py]: done

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_database.py [hunk 31]
```diff
@@ -0,0 +1,368 @@
...
+        "_{}_".format(random_string(6)),
+        random_string(6, digit=True),
+        random_string(6).upper(),
+    ])
+    def test_exists(self, db_name):
+        '''
+        @summary: Testing exist method of Database API
+        '''
+        if len(db_name) >= 128 and (self.liteserv_platform != "ios" or self.liteserv_platform != "android"):
+            pytest.skip("Test not supported on .net platfomrs")
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/CBL_Functional_tests/APITests/test_database.py]: done

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_database.py [hunk 32]
```diff
@@ -0,0 +1,368 @@
...
+        random_string(6, digit=True),
+        random_string(6).capitalize(),
+        random_string(6).upper(),
+    ])
+    def test_getName(self, db_name):
+        '''
+        @summary: Testing getName method of Database API
+        '''
+        if len(db_name) >= 128 and (self.liteserv_platform != "ios" or self.liteserv_platform != "android"):
+            pytest.skip("Test not supported on .net platfomrs")
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/CBL_Functional_tests/APITests/test_database.py]: done

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_database.py [hunk 33]
```diff
@@ -0,0 +1,368 @@
...
+        '''
+        doc_id_prefix = "bar"
+        doc1 = self.doc_obj.create(doc_id=doc_id_prefix)
+        doc2 = self.doc_obj.create(doc_id=doc_id_prefix)
+        db_1 = self.db_obj.create(db1)
+        db_2 = self.db_obj.create(db2)
+        self.db_obj.saveDocument(db_1, doc1)
+        self.db_obj.saveDocument(db_2, doc2)
+        self.db_obj.purge(document=doc1, database=db_1)
+        # assert self.db_obj.getDocument(db_2, doc_id) == -1
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/CBL_Functional_tests/APITests/test_database.py]: done

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_database.py [hunk 34]
```diff
@@ -0,0 +1,368 @@
...
+        (random_string(6, digit=True), random_string(6)),
+        (random_string(6).capitalize(), random_string(6)),
+        (random_string(6).upper(), random_string(6))
+    ])
+    def test_saveDocument(self, db_name, doc_id):
+        '''
+        @summary: Testing save method of Database API
+        '''
+        if len(db_name) >= 128 and (self.liteserv_platform != "ios" or self.liteserv_platform != "android"):
+            pytest.skip("Test not supported on .net platfomrs")
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/CBL_Functional_tests/APITests/test_database.py]: done

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_dictionary.py [hunk 35]
```diff
@@ -0,0 +1,265 @@
...
+        '''
+        content_dict = self.dict_obj.create()
+        self.dict_obj.setBoolean(content_dict, key, value)
+        assert self.dict_obj.getBoolean(content_dict, key) == value
+
+    def test_get_set_date(self):
+        '''
+        @summary: Testing get and set Date methods of Dictionary API
+        '''
+        # TODO implementation does not work on ios
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/CBL_Functional_tests/APITests/test_dictionary.py]: removed todo comment

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_dictionary.py [hunk 36]
```diff
@@ -0,0 +1,265 @@
...
+        hashmap[key] = value
+        content = self.dict_obj.create(hashmap)
+        self.dict_obj.remove(content, key)
+        assert not self.dict_obj.contains(content, key)
+
+    def test_toMap(self):
+        '''
+        @summary: Testing remove method of Dictionary API
+        '''
+        # TODO ios gets back {} for self.datatype.hashMap()
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/CBL_Functional_tests/APITests/test_dictionary.py]: yes and removed todo comment.

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_document.py [hunk 37]
```diff
@@ -0,0 +1,348 @@
...
+        (random_string(6), "{}".format(random.randint(10, 99))),
+        (random_string(6), "{}".format(random.randint(100, 999))),
+        (random_string(6), "{}".format(random.randint(1000, 9999)))
+    ])
+    def test_get_set_long(self, key, value):
+        '''
+        @summary: Testing Get and Set Float method of Document API
+        '''
+        doc = self.doc_obj.create()
+        # long_obj = self.datatype.setLong(value)
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/CBL_Functional_tests/APITests/test_document.py]: done

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_document.py [hunk 38]
```diff
@@ -0,0 +1,348 @@
...
+        doc1b_doc = self.db_obj.getDocument(db, doc1b_id)
+        doc1b_todict = self.doc_obj.toMap(doc1b_doc)
+
+        expected_dict = {
+            "dict1": {"name": "n1"},
+            "dict1b": {"name": "n1"},
+            "dict2": {"name": "n2"}
+        }
+
+        assert sorted(expected_dict) == sorted(doc1b_todict)
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/CBL_Functional_tests/APITests/test_document.py]: Done

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_document.py [hunk 39]
```diff
@@ -0,0 +1,348 @@
...
+        doc1b_doc = self.db_obj.getDocument(db, doc1b_id)
+        doc1b_todict = self.doc_obj.toMap(doc1b_doc)
+
+        expected_dict = {
+            "array1": ["a1", {"name": "n1"}],
+            "array1b": ["a1", {"name": "n1"}],
+            "array2": ["a2", {"name": "n2"}]
+        }
+
+        assert sorted(expected_dict) == sorted(doc1b_todict)
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/CBL_Functional_tests/APITests/test_document.py]: done

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py (hunk 20 above)
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: changed it to log_info

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py (hunk 21 above)
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: changed it to log_info

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py (hunk 22 above)
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: changed it to log_info

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py (hunk 23 above)
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: actually, we are calling this method in parallel for all devices as the method waits for the replication to become idle on all devices. That's why I'm using thread.

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py (hunk 24 above)
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: changed it to log_info

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py (hunk 25 above)
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: It is intended to achieve parallelism

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py (hunk 26 above)
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: answered above

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py (hunk 27 above)
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: changed it to log_info

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py (hunk 28 above)
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: changed it to log_info

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py (hunk 19 above)
Comment (sridevi-15) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: comment still exists, if not needed, better to remove it

**Comment by hrajput89** [comment]
Closing this and creating a new PR

**Pull request closed without merging**